test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/NullPointersTest.scala

```diff
@@ -0,0 +1,289 @@
+
+    assertNPE(nullOf[Tester].inlineMethodWithField(1))
+    assertNPE(inlineNullOf[Tester].inlineMethodWithField(1))
+
+    assertNPE(nullOf[Tester].toString())
+    assertNPE(inlineNullOf[Tester].toString())
+    assertNPE(nullOf[AnyRef].toString())
+    assertNPE(inlineNullOf[AnyRef].toString())
+
+    assertNPE(nullOf[Tester].synchronized("foo"))
+    assertNPE(inlineNullOf[Tester].synchronized("foo"))
+    assertNPE(nullOf[AnyRef].synchronized("foo"))
+    assertNPE(inlineNullOf[AnyRef].synchronized("foo"))
+
+    assertNPE(nullOf[Tester].getClass())
+    assertNPE(inlineNullOf[Tester].getClass())
+    assertNPE(nullOf[AnyRef].getClass())
+    assertNPE(inlineNullOf[AnyRef].getClass())
+
+    assertNPE(nullOf[Tester].clone())
+    assertNPE(inlineNullOf[Tester].clone())
+
+    // NPE takes precedence over evaluating arguments, unlike on the JVM
+
+    if (executingInJVM) {
+      assertThrows(classOf[IllegalArgumentException],
+          nullOf[Tester].noInlineMethod(throwIllegalArgAsInt()))
+    } else {
+      assertNPE(nullOf[Tester].noInlineMethod(throwIllegalArgAsInt()))
+      assertNPE(nullOf[Tester].noInlineMethod(throwIllegalArgAsIntInline()))
+      assertNPE(inlineNullOf[Tester].noInlineMethod(throwIllegalArgAsInt()))
+      assertNPE(inlineNullOf[Tester].noInlineMethod(throwIllegalArgAsIntInline()))
+
+      assertNPE(nullOf[Tester].inlineMethod(throwIllegalArgAsInt()))
+      assertNPE(nullOf[Tester].inlineMethod(throwIllegalArgAsIntInline()))
+      assertNPE(inlineNullOf[Tester].inlineMethod(throwIllegalArgAsInt()))
+      assertNPE(inlineNullOf[Tester].inlineMethod(throwIllegalArgAsIntInline()))
+    }
+  }
+
+  @Test def methodCallsWithHijackedClasses(): Unit = {
+    assumeCompliantNPEs()
+
+    assertNPE(nullOf[Integer].intValue())
+    assertNPE(inlineNullOf[Integer].intValue())
+    assertNPE(nullOf[Integer].toString())
+    assertNPE(inlineNullOf[Integer].toString())
+    assertNPE(nullOf[Integer].synchronized("foo"))
+    assertNPE(inlineNullOf[Integer].synchronized("foo"))
+    assertNPE(nullOf[Integer].getClass())
+    assertNPE(inlineNullOf[Integer].getClass())
+
+    assertNPE(nullOf[Character].charValue())
+    assertNPE(inlineNullOf[Character].charValue())
+    assertNPE(nullOf[Character].toString())
+    assertNPE(inlineNullOf[Character].toString())
+
+    assertNPE(nullOf[String].length())
+    assertNPE(inlineNullOf[String].length())
+    assertNPE(nullOf[String].charAt(3))
+    assertNPE(inlineNullOf[String].charAt(3))
+    assertNPE(nullOf[String].toString())
+    assertNPE(inlineNullOf[String].toString())
+    assertNPE(nullOf[String].concat("foo"))
+    assertNPE(inlineNullOf[String].concat("foo"))
+
+    // NPE takes precedence over evaluating arguments, unlike on the JVM
+
+    if (executingInJVM) {
+      assertThrows(classOf[IllegalArgumentException],
+          nullOf[String].substring(throwIllegalArgAsInt()))
+    } else {
+      // charAt is a bit special
```

Comment:
Maybe clarify that the implementation is special, not the behavior. I tripped over this.